gradio/route_utils.py

```diff
@@ -0,0 +1,249 @@
+import json
+from copy import deepcopy
+from typing import TYPE_CHECKING, Dict, List, Optional, Union
+
+import fastapi
+from gradio_client.documentation import document, set_documentation_group
+
+from gradio import utils
+from gradio.data_classes import PredictBody
+from gradio.exceptions import Error
+from gradio.helpers import EventData
+
+if TYPE_CHECKING:
+    from gradio.routes import App
+
+set_documentation_group("routes")
+
+
+class Obj:
+    """
+    Using a class to convert dictionaries into objects. Used by the `Request` class.
+    Credit: https://www.geeksforgeeks.org/convert-nested-python-dictionary-to-object/
+    """
+
+    def __init__(self, dict_):
+        self.__dict__.update(dict_)
+        for key, value in dict_.items():
+            if isinstance(value, (dict, list)):
+                value = Obj(value)
+            setattr(self, key, value)
+
+    def __getitem__(self, item):
+        return self.__dict__[item]
+
+    def __setitem__(self, item, value):
+        self.__dict__[item] = value
+
+    def __iter__(self):
+        for key, value in self.__dict__.items():
+            if isinstance(value, Obj):
+                yield (key, dict(value))
+            else:
+                yield (key, value)
+
+    def __contains__(self, item) -> bool:
+        if item in self.__dict__:
+            return True
+        for value in self.__dict__.values():
+            if isinstance(value, Obj) and item in value:
+                return True
+        return False
+
+    def keys(self):
+        return self.__dict__.keys()
+
+    def values(self):
+        return self.__dict__.values()
+
+    def items(self):
+        return self.__dict__.items()
+
+    def __str__(self) -> str:
+        return str(self.__dict__)
+
+    def __repr__(self) -> str:
+        return str(self.__dict__)
+
+
+@document()
+class Request:
+    """
+    A Gradio request object that can be used to access the request headers, cookies,
+    query parameters and other information about the request from within the prediction
+    function. The class is a thin wrapper around the fastapi.Request class. Attributes
+    of this class include: `headers`, `client`, `query_params`, and `path_params`. If
+    auth is enabled, the `username` attribute can be used to get the logged in user.
+    Example:
+        import gradio as gr
+        def echo(name, request: gr.Request):
+            print("Request headers dictionary:", request.headers)
+            print("IP address:", request.client.host)
+            return name
+        io = gr.Interface(echo, "textbox", "textbox").launch()
+    """
+
+    def __init__(
+        self,
+        request: Union[fastapi.Request, None] = None,
```

Comment:
This PEP604 syntax (`X | Y` for `Union[X, Y]`) is not supported by Python<3.10 and the following error has been raised in the CI on this branch actually,
but looks like the original code (`routes.py`) has passed the tests on the main branch... why?
```
No configuration file found.
pyproject.toml file found at /home/runner/work/gradio/gradio.
Loading pyproject.toml file at /home/runner/work/gradio/gradio/pyproject.toml
Assuming Python version 3.8
Assuming Python platform Linux
Searching for source files
Found 79 source files
pyright 1.1.305
/home/runner/work/gradio/gradio/gradio/route_utils.py
  /home/runner/work/gradio/gradio/gradio/route_utils.py:89:23 - error: Alternative syntax for unions requires Python 3.10 or newer
  /home/runner/work/gradio/gradio/gradio/route_utils.py:88:34 - error: Alternative syntax for unions requires Python 3.10 or newer
  /home/runner/work/gradio/gradio/gradio/route_utils.py:[19](https://github.com/gradio-app/gradio/actions/runs/5996574137/job/16261357539?pr=5165#step:14:20)2:[25](https://github.com/gradio-app/gradio/actions/runs/5996574137/job/16261357539?pr=5165#step:14:26) - error: Alternative syntax for unions requires Python 3.10 or newer
3 errors, 0 warnings, 0 informations
```

Reply:
You just need to add:

```py
from __future__ import annotations
```

at the top of your file

Reply:
omg! Thanks :D